h261_play.1
.\" @(#) $Header$ (LBL)
.\"
.\" Copyright (c) 1994-1995
.\" The Regents of the University of California.  
.\" All rights reserved.  
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that: (1) source code distributions
.\" retain the above copyright notice and this paragraph in its entirety, (2)
.\" distributions including binary code include the above copyright notice and
.\" this paragraph in its entirety in the documentation or other materials
.\" provided with the distribution, and (3) all advertising materials mentioning
.\" features or use of this software display the following acknowledgment:
.\" ``This product includes software developed by the University of California,
.\" Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
.\" the University nor the names of its contributors may be used to endorse
.\" or promote products derived from this software without specific prior
.\" written permission.  
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  
.\"
.TH H261_PLAY 1 "04 Apr 1996"
.de HD
.sp 1.5
.B
..
.SH NAME
h261_play \- view and dump h261 files
.SH SYNOPSIS
.na
.B h261_play
[
.B \-b
]
[
.B \-C
]
[
.B \-c
.I dither
]
[
.B \-D
]
[
.B \-d
.I display
]
[
.B \-l
]
[
.B \-n
]
[
.B \-p
]
[
.B \-S
]
[
.B \-s
]
[
.B \-t
]
[
.B \-u
.I script
]
[
.B \-V
.I visual
]
.I file
.br
.ad
.SH DESCRIPTION
.LP
.I H261_play
is a quick-and-dirty diagnostic we hacked together for debugging
our h261 decoder.  The input file should be a raw h261 bit streams
(i.e., the bit stream specified in ITU recommendation H.261 minus
the framing and error correction coding from fig. 13 and sec. 5.4).
H261_play will display this bit stream and, optionally, provide a
detailed diagnostic dump of it to stdout.  Keyboard commands are
available to single step and get some interactive diagnostic output.
.SH OPTIONS
.TP
.B \-b
make the rendering window `big' (2 x CIF).  Other window size flags are
.B \-s
(small or QCIF) and
.B \-t
(tiny or 1/8 CIF).  If none of these flags is given, the window is
made CIF size.  The h261 image is scaled to to the window size.
.TP
.B \-c
On a color-mapped display, use the algorithm indicated by
.I dither
to convert video (typically represented in 24-bit YUV color space)
to the available color palette.  For monochrome and truecolor
displays, this argument is ignored.
.I dither
may be one of the following:
.nf
	ed		error-diffusion (i.e., Floyd-Steinberg) dither 
	gray		32-levels of gray
	od		a simple 4x4 ordered dither
	quantize		straight quantization
.fi
The ``od'' algorithms uses the standard 5x5x5 color cube.
Since this color palette is used by several other applications
(wb, nv, and gs), colors can be shared which avoids the use of
a private color map.
.TP
.B \-C
show changes.  The macroblocks that changed in the current frame
are display normally while the blocks that didn't change are diplayed
in a washed-out gray.  (This option is generally only useful when
you are single stepping through the file.)
.TP
.B \-d
Connect to the X server indicated by the
.I display
argument.
.TP
.B \-D
debug \- send a very verbose, block-by-block dump of the h.261
data stream to stdout.  (See below for format.)
.TP
.B \-l
`loop' \- when the input file eof is reached, it is rewound
and restarted.
.TP
.B \-n
`no rendering' \- the input file is read and decoded (and dumped
if
.B \-D
is specified) but nothing is output to the screen.  This
flag can be used to get an ascii dump of an h261 file or
to time the decoder.
.TP
.B \-p
.I Don't
pause on startup.  Normal behavior is to start up in pause mode
so a space or carriage return is required to start decode and
rendering.
.TP
.B \-s
render into a `small' (QCIF) window.
.TP
.B \-S
Don't use the X shared memory extension to convey video images
to the X server.
.TP
.B \-t
render into a `tiny' (1/8 CIF) window.
.TP
.B \-u
Source
.I script,
in addition to the compiled-in script, to build the user interface.
This is normally only useful during development.
.TP
.B \-V
Use
.I visual
as the rendering window visual.

.SH "KEYBOARD COMMANDS"
The following can be typed in the h261_play video window:
.TP
.B space
toggles run/pause.
.TP
.B "carriage return"
if paused, single steps one frame forward.
.TP
.B q
quit \- exits the program.
.TP
.B d
dump \- dumps the current frame as a PPM file to `t.ppm'
.TP
.B f
frame \- writes the current frame number on stdout.
.TP
.B "button 1"
writes the gob and macroblock number of the data under the cursor
to stdout.

.SH "DEBUGGING OUTPUT"
The debugging output consists of one line for each syntactic unit
in the h.261 file.  Most of the output should be self explanatory,
given a copy of the H.261 spec.  Two things might not be are:
.TP
    \(bu
Each line ends with a representation of the bits that make up this
syntactic unit in the form `cnt/bstring' where
.I cnt
is the number of bits in the unit and
.I bstring
is the actual bit string in hex, right adjusted with 0 fill to a multiple
of 4 bits on the left.
.TP
    \(bu
In image coef blocks (output lines that start with
.IR blk ),
all non-zero coefs are listed in the form `number:value' where
.I number
is the coef number in zig-zag order with 0 being DC and
.I value
is the 
.I dequantized
coef value.

.SH "SEE ALSO"
vic(1),
ppmtolut(1),
mkcube(1),
histtolut(1)
.LP
McCanne, Steven and Jacobson, Van.
``vic: A Flexible Framework for Packet Video''.
In proceedings of ACM Multimedia '95.
November, 1995.
.br
.SH ACKNOWLEDGMENTS
.LP
Elan Amir (elan@cs.berkeley.edu) implemented the error diffusion
dithering code and dynamic color allocation (median cut) algorithms.
Chris Goodman (goodman@sgi.com) provided valuable advice on the
error diffusion algorithm and helped debug the implementation.
.LP
Martin Vetterli (martin@diva.eecs.berkeley.edu) provided input on fast
DCT implementations.  He pointed out that Arai, Agui, and Nakajmia's
8pt 1D DCT can be used to compute scaled row and column DCTs leading to
a 80 multiply 8x8 2D DCT.
.LP
Thanks to Robert Berger (rwb@J.GP.CS.CMU.EDU) for his excellent
web page on monitor gamma and for his permission to redistribute the
gamma calibration test image (gamma.gif).
.LP
This software is based in part on the work of the Independent JPEG Group
and the Portable Video Research Group.
.LP
This work was co-sponsored by the the Lawrence Berkeley National Laboratory
and the Tenet Group of the University of California Berkeley and 
of the International Computer Science Institute.
Support was provided by
(1) an AT&T Graduate Fellowship;
(2) for Lawrence Berkeley National Laboratory: (i) the Director,
Office of Energy Research, Scientific Computing Staff, of the
U.S. Department of Energy, Contract No. DE-AC03-76SF00098,
(ii) Sun Microsystems, (iii) Digital Equipment Corporation,
and (iv) Silicon Graphics Inc.; and
(3) for the Tenet Research Group: (i) the National Science Foundation
and the Advanced Research Projects Agency (ARPA) under
Cooperative Agreement NCR-8919038 with the Corporation for
National Research Initiatives, (ii) Digital Equipment Corporation,
and (iii) Silicon Graphics Inc.
.LP
.SH AUTHOR
Steven McCanne (mccanne@ee.lbl.gov), University of California, Berkeley
and Lawrence Berkeley National Laboratory, Berkeley, CA, and
Van Jacobson (van@ee.lbl.gov),
Lawrence Berkeley National Laboratory, Berkeley, CA.
.SH BUGS
There are no contrast or brightness controls.

The error-diffsuion dithering code needs more work.
At low luminosities, strange pastel colors appear.
Blue skies are often rendered green.

It would be nice to single-step in smaller units than a frame.


tmn/CHANGES
		  Changes from version 3.0 to 3.1
                  ===============================

Encoder & Decoder:

	Changes from H.263+ Draft 20 are incorporated.

	H.263+ Annex D (unrestricted motion vector) with RVLC is
	implemented. H.263 Annex D can still be used (with option -D
	1) to avoid PLUSPTYPE. 

	Annex O Spatial (1- and 2-dimensional) and SNR scalability are
	now supported. 

        Custom picture format (factor of 16, pixel aspect ratio 12:11
        only) is supported.  

	Bug fixes:  Bugs (including mode combination problems)
        reported in version 3.0 were fixed. 
 
Encoder:

	TMN-8 rate control is implemented (TMN-5 rate control also available)

        Full search motion estimation is available through
        pre-processor definition (sim.h). Fast search is the default
        method. 

Decoder:

	TMN-8 Post Filter implemented

See Help (tmn -h) to enable modes.


		  Changes from version 2.0 to 3.0
		  ===============================


Department of Electrical Engineering, University of British Columbia, 
CANADA (http:/www.ee.ubc.ca/image/)  is now the official maintainer of
the Public Domain H.263+ encoder/decoder.

Our H.263+ Coder supports the following H.263+ Annexes (All H.263
Annexes are still supported), based on H.263+ draft 13 document
:

Annex I:    Advanced Intra Coding mode
Annex J:    Deblocking Filter mode
Annex M:    Improved PB-frames mode
Annex O:    Temporal (only) Scalability mode
Annex S:    Alternative Inter VLC mode
Annex T:    Modified Quantization mode

This is the first release of this code and might not run without error
when combining options.  Please report bugs to guyc@ee.ubc.ca.

Encoder:
*	New fast search motion estimation technique of TMN-8 is used.  New technique reduces
	encoding time from 3:1 compared to  
	Full Search with an average loss of less than 0.01 Db in PSNR.



		   Changes from version 1.7 to 2.0
		   ===============================

Encoder & Decoder:
*	Because of the possible patent infringements in all
	implementations of H.263, I decided to remove the
        GNU License introduced in version 1.7.
	The software is now distributed under the same
        "disclaimer of warranty" as the freely available 
	MPEG-2 software. See the file DISCLAIM for details.

Encoder:
*	Usage is printed to stdout instead of stderr to 
	enable piping to file on Win95 systems.	


		   Changes from version 1.6c to 1.7
		   ================================

Encoder & Decoder:
*	Changed the copyright to put all the software under the GNU
	General Public License. Permission has been granted from all
	parties that have contributed to this version of the software.

*	It is no longer necessary to compile separate versions
	for encoding and decoding the different coding formats
	(SQCIF, QCIF, CIF, 4CIF, 16CIF)

Encoder:
*	Removed all code coming from the Stanford PVRG group. This was
	the bitstream functions and the VLC encoding. New bitstream
	functions and new VLC-tables were written based the idea
	from the MPEG-2 simulation group software.

*	Added back the low-delay, variable frame rate, rate control 
	from version 1.5.
	To use the rate control from version 1.6, compile with
	OFFLINE_RATE_CONTROL defined. Read more about the rate 
	control methods in ratectrl.c.


Decoder:
*	Display for Windows 95/NT has been added. This works much
	the same way as the X11 display: when the decoder is
	run in a console window, a window displaying the decoded
	frames pops up.

*	Variable frame rate display has been added. In this
	mode the decoder will read the frame rate from the bitstream.
	This is now the default display mode.

*	When decoding PB-frames the display rate should
	be reasonably correct, provided the computer 
	running the decoder is fast enough.


		   Changes from version 1.6b to 1.6c
		   =================================

Encoder & Decoder:
*	Fixed a bug in OBMC prediction for the P-frame in PB-mode 
	when surrounding block is intra encoded. 

Decoder:
*	Fixed a bug in packed 24-bit display mode.


		   Changes from version 1.6a to 1.6b
		   =================================

Encoder & Decoder:
*	Fixed a bug in OBMC prediction when the surrounding block
	is intra encoded.

Encoder:
*	Fixed Temporal Reference in the PB-frames mode.
	Changed some variable names related to frame skipping
	to make them more understandable.

*	Fixed a problem when Unrestricted MV mode was used
	together with several GOB syncs per picture.

Decoder:
*	Is now able to handle stuffing in SAC mode (I hope)



		   Changes from version 1.6 to 1.6a
		   ================================

Decoder:
*	Support for display at 16 and 24 (32) bits colordepths. 
	No dithering is used in these mode. Display looks
	great, but speed is slower than in 8 bits mode.



		   Changes from version 1.5 to 1.6
		   ===============================

Encoder & Decoder:
*	The increased motion vector range has been included 
	in the Unrestricted Motion Vector mode. This increases
	performance considerably for sequences with large
	amounts of camera motion. Note that in the Advanced
	Prediction mode, motion vectors are still allowed to
	point outside the picture, but the UMV mode is not
	automatically turned on.

*	Other changes in the final version of H.263:
	+ back to old half pel interpolation scheme
	+ calculation of chroma vectors in PB mode

*	Bug-fixes and code clean-up:
	+ most reported problems with version 1.5 
	+ several others, but none essential for coding
	  performance
	

Encoder:
*	When using PB-frames the adaptive switch which turns
	off PB-frames when they are not effective, has been
	included. This is always used in the PB-frames mode, 
	and encodes the next frames as two P-pictures instead 
	of a PB-frame when this is more effective.

*	The TMN5 rate control has been replaced by a simplified 
	rate control with fixed frame rate and a quantizer 
	change for each picture. Read the comments in 
	ratectrl.c for more information.

	(both algorithms above were used when generating

rtp/h261_play.cpp
/*
 * Copyright (c) 1994 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and the Network Research Group at
 *      Lawrence Berkeley Laboratory.
 * 4. Neither the name of the University nor of the Laboratory may be used
 *    to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
static const char rcsid[] =
    "@(#) $Header$ (LBL)";

#include <stdio.h>
#ifdef WIN32
#include <winsock.h>
#include "config.h"
#else
#include <sys/param.h>
#include <sys/time.h>
#include <netinet/in.h>
#endif
#ifdef USE_SHM
extern "C" {
#include <sys/ipc.h>
#include <sys/shm.h>
#if defined(sparc) && !defined(__svr4__)
int shmget(key_t, int, int);
char *shmat(int, char*, int);
int shmdt(char*);
int shmctl(int, int, struct shmid_ds*);
#endif
}
#endif

#ifdef WIN32
#include <stdlib.h>
//#else
//#include <osfcn.h>
#endif

#include "p64.h"
#include "p64dump.h"
#include "color.h"
#include "renderer.h"
#include "vw.h"
#include "vic_tcl.h"

void
usage()
{
	fprintf(stderr, "usage\n");
	exit(1);
}

#ifndef USE_SHM
int use_shm = 0;
#else
int use_shm = 1;
static int
noXShm(Display* , XErrorEvent*)
{
	use_shm = 0;
	fprintf(stderr, "h261_play: warning: not using shared memory\n");
	return (0);
}

/*
 * XXX this is a hack to see if we can used shared memory.
 * if the X server says we can, and we\'re on the same
 * host as the X server, then we are golden.
 */
static void
checkXShm(Display* dpy, const char* )
{
	XShmSegmentInfo si;
	if ((si.shmid = shmget(IPC_PRIVATE, 512, IPC_CREAT|0777)) < 0) {
		perror("shmget");
		exit(1);
	}
	si.readOnly = 1;
	XSync(dpy, 0);
	XSetErrorHandler(noXShm);
	XShmAttach(dpy, &si);
	XSync(dpy, 0);
	XSetErrorHandler(NULL);
	(void)shmctl(si.shmid, IPC_RMID, 0);
}
#endif

P64Decoder* decoder;
Renderer* renderer;
FILE* in;

int nflag = 0;

int nframe = 0;
timeval start;

struct offset {
	off_t off;
	int bit;
};
#define MAXFRAME 100000
offset offs[MAXFRAME];

void
dump_ppm(char* file, const u_char* yp, const u_char* up, const u_char* vp,
	 int w, int h)
{
	FILE *f;
	if (file == 0)
		f = stdout;
	else {
		(void)unlink(file);
		f = fopen(file, "w");
		if (f == 0)
			/*XXX*/
			return;
	}
	fprintf(f, "P3\n%d %d\n255\n", w, h);
	for (int y = 0; y < h; ++y) {
		for (int x = 0; x < w; ++x) {
			double Y = (double)yp[w * y + x];
			double U = up[(w / 2) * (y / 2) + x / 2] - 128.;
			double V = vp[(w / 2) * (y / 2) + x / 2] - 128.;
			
			int r = int(Y + 1.402 * V);
			int g = int(Y - 0.34414 * U - 0.71414 * V);
			int b = int(Y + 1.772 * U);
#define LIM(x) ((x) >= 255 ? 255 : ((x) < 0 ? 0 : (x)))
			r = LIM(r);
			g = LIM(g);
			b = LIM(b);
			fprintf(f, "%d %d %d\n", r, g, b);
		}
	}
	if (f == stdout)
		fflush(stdout);
	else
		fclose(f);
}

void
buildindex(FILE* f)
{
	off_t off = 0;
	int n = 0;
	u_int bb;

	bb = getc(f);
	bb <<= 8;
	bb |= getc(f);
	int v = getc(f);
	bb <<= 4;
	bb |= v >> 4;
	v <<= 4;
	for (;;) {
		int bit;
		for (bit = 0; bit < 4; ++bit) {
			if ((bb & 0xfffff) == 0x00010) {
				offset* p = &offs[n++];
				p->bit = bit;
				p->off = off;
			}
			bb <<= 1;
			v <<= 1;
			bb |= (v >> 8) & 1;
		}
		v = getc(f);
		if (feof(f))
			break;
		for (; bit < 8; ++bit) {
			if ((bb & 0xfffff) == 0x00010) {
				offset* p = &offs[n++];
				p->bit = bit;
				p->off = off;
			}
			bb <<= 1;
			v <<= 1;
			bb |= (v >> 8) & 1;
		}
		++off;
	}
	/*XXX skip last frame, since we use offs[cfrm+1] below*/
	nframe = n - 1;
	rewind(f);
}

int lflag;
int cfrm;

/*XXX*/
u_char crvec[768 * 576 / 64];
u_int now = 1;

void render_frame(const u_char* frm, int w, int h)
{
	/*
	 * Go through all the timestamps and smash the time that
	 * is about to wrap from the past into the future to the present,
	 * so that the block gets updated just once.  If we let timestamps
	 * wrap without doing anything, the algorithm would still work,
	 * but we'd end up rendering a stationary block on every call here
	 * (until we surpassed now + 128 again).  In other words, for
	 * an unchanging block, this approach renders it 2 times out of 256,
	 * rather than 128 out of 256.
	 */
	int wraptime = now ^ 0x80;
	u_char* ts = crvec;
	for (int k = sizeof(crvec); --k >= 0; ++ts) {
		if (*ts == wraptime)
			*ts = now;
	}

	YuvFrame f(now, (u_int8_t*)frm, crvec, w, h);
	renderer->consume(&f);
	now = (now + 1) & 0xff;
	decoder->mark(now);
}

int nextframe()
{
	if (cfrm >= nframe-1) {
		if (!lflag) {
			fprintf(stderr, "h261_play: eof\n");
			timeval end;
			gettimeofday(&end, 0);
			double v = end.tv_sec - start.tv_sec;
			v += 1e-6 * (end.tv_usec - start.tv_usec);
			fprintf(stderr,
			"h261_play: %d frames in %g seconds (%g fps)\n",
				nframe, v, (double)nframe / v);
			exit(0);
		}
		cfrm = 0;
	}
	if (cfrm == 0)
		gettimeofday(&start, 0);

	offset* p = &offs[cfrm++];
	offset* nxt = &offs[cfrm];

	if (fseek(in, p->off, 0) < 0) {
		perror("fseek");
		exit(1);
	}
	int len = nxt->off - p->off + 1;
	int ebit = 8 - nxt->bit;
	if (ebit == 8) {
		len -= 1;
		ebit = 0;
	}

	u_char buffer[80*1024];
	if (fread((char*)buffer, len, 1, in) == 0) {
		perror("fread");
		exit(1);
	}

	P64Decoder* d = decoder;
	d->decode(buffer, len, p->bit, ebit, 0, 0, 0, 0, 0);
	d->sync();
	if (nflag)
		return (TCL_OK);

	render_frame(d->frame(), d->width(), d->height());
	return (TCL_OK);
}

extern "C" char *optarg;
extern "C" int optind;
extern "C" int opterr;

const char*
disparg(int argc, const char** argv, const char* optstr)
{
	const char* display = 0;
	int op;
	while ((op = getopt(argc, (char**)argv, (char*)optstr)) != -1) {
		if (op == 'd') {
			display = optarg;
			break;
		}
		else if (op == '?')
			usage();
	}
	optind = 1;
	return (display);
}

/*
 * decoder decode $frmno
 * decoder dump [file]
 * decoder next
 * decoder frame
 */
int
cmd_decoder(ClientData , Tcl_Interp*, int argc, const char** argv)
{
	Tcl& tcl = Tcl::instance();
	if (argc == 2) {
		if (strcmp(argv[1], "frame") == 0) {
			char* bp = tcl.buffer();
			sprintf(bp, "%d", cfrm);
			tcl.result(bp);
			return (TCL_OK);
		}
		if (strcmp(argv[1], "next") == 0) {
			nextframe();
			return (TCL_OK);
		}
	} else if (argc == 3) {
		if (strcmp(argv[1], "renderer") == 0) {
			renderer = (Renderer*)TclObject::lookup(argv[2]);
			return (TCL_OK);
		}
#ifdef notdef
		if (strcmp(argv[1], "dump") == 0) {
			dump_ppm(argv[2],
				 decoder->yplane(),
				 decoder->uplane(),
				 decoder->vplane(), 
				 decoder->width(), decoder->height());
			return (TCL_OK);
		}
#endif
	}
	return (TCL_ERROR);
}

#ifndef WIN32
extern "C" {
int
TkPlatformInit(Tcl_Interp *interp)
{
	extern void TkCreateXEventSource(void);
	Tcl_SetVar(interp, "tk_library", "", TCL_GLOBAL_ONLY);
	TkCreateXEventSource();
	return (TCL_OK);
}
}
#endif

int
main(int argc, const char **argv)
{
	int op;
	int Dflag = 0;
	int osize = 2;

	const char* optstr = "bCc:Dd:lnpSstu:V:";
	const char* display = disparg(argc, argv, optstr);
	Tcl::init("h261_play");
	Tcl& tcl = Tcl::instance();
#ifdef WIN32
	if (display == NULL)
		display = "localhost:0";
#endif
	tcl.evalf(display?
		    "set argv \"-name h261_play -display %s\"" :
		    "set argv \"-name h261_play\"",
		  display);
	Tk_Window tk = 0;
	if (Tk_Init(tcl.interp()) == TCL_OK)
		tk = Tk_MainWindow(tcl.interp());
	if (tk == 0) {
		fprintf(stderr, "h261_play: %s\n", tcl.result());
		exit(1);
	}
	tcl.tkmain(tk);
	EmbeddedTcl::init();
	tcl.evalc("init_resources");

	while ((op = getopt(argc, (char**)argv, optstr)) != -1) {
		switch (op) {

		default:
			usage();

		case 'b':
			osize = 3;
			break;

		case 'C': 
			tcl.add_option("showChanges", "true");
			break; 

		case 'c':
			tcl.add_option("dither", optarg);
			break;

		case 'D': 
			++Dflag;
			break; 

		case 'd':
			break;

		case 'l':
			++lflag;
			break;

		case 'n':
			++nflag;
			break;

		case 'p':
			tcl.add_option("pause", "false");
			break;

		case 's':
			osize = 1;
			break;

		case 'S':
			use_shm = 0;
			break;

		case 't':
			osize = 0;
			break;

		case 'u':
			tcl.add_option("startupScript", optarg);
			break;

		case 'V':
			tcl.add_option("visual", optarg);
			break;

		}
	}
#ifdef USE_SHM
	if (use_shm)
		checkXShm(Tk_Display(tk), display);
#endif

	if (optind >= argc || argc <= 1)
		usage();
	const char* infile = argv[optind];

	if (Dflag)
		decoder = new P64Dumper;
	else
		decoder = new FullP64Decoder;

	decoder->marks(crvec);
	decoder->mark(now);

#ifdef WIN32
	in = fopen(infile, "rb");
#else
	in = fopen(infile, "r");
#endif
	if (in == 0) {
		perror(infile);
		exit(1);
	}
	/*
	 * Check that the file begins with a picture header.
	 */
	u_char phdr[3];
	if (fread(phdr, 3, 1, in) != 1 ||
	    phdr[0] != 0 || phdr[1] != 1 || (phdr[2] >> 4) != 0) {
		fprintf(stderr, "h261_play: %s not a raw H.261 bit stream\n",
			infile);
		exit(1);
	}
	rewind(in);
	fprintf(stderr, "h261_play: building frame index...");
	buildindex(in);
	fprintf(stderr, "ready - %d frames\n", nframe);

	tcl.CreateCommand("decoder", cmd_decoder);
	tcl.evalf("h261_main %d %d", 88 << osize, 72 << osize);

	Tk_MainLoop();
	exit(0);
}
